frontend/src/test/scala/bloop/engine/FileWatchingSpec.scala

```diff
@@ -160,4 +166,39 @@ class FileWatchingSpec {
 
     testFileWatcher(state, testProject)(workerAction, testAction)
   }
+
+  @Test
+  def cancelFileWatching(): Unit = {
+    def simulation(state: State): Task[State] = Task {
+      println("Simulating compilation...")
+      Thread.sleep(1000)
+      println("Simulating completed.")
+      state
+    }
+
+    val javaEnv = JavaEnv.default
+    val instance = CompilationHelpers.scalaInstance
+    val structures = Map(RootProject -> Map("A.scala" -> ArtificialSources.`A.scala`))
+    withState(structures, Map.empty, scalaInstance = instance, javaEnv = javaEnv) {
+      (state: State) =>
+        val rootProject = state.build.getProjectFor(RootProject).get
+        val watchTask = Interpreter.watch(rootProject, state, simulation _)
+        val s1 = Scheduler.computation(parallelism = 2)
+        val handle = watchTask
+          .executeWithOptions(_.enableAutoCancelableRunLoops)
+          .runAsync(state.scheduler)
+
+        val waitForWatching = Task(Await.result(handle, Duration.Inf))
+        val waitAndCancelWatching =
+          Task(handle.cancel()).delayExecution(FiniteDuration(2, TimeUnit.SECONDS))
+        val waitWatchingHandle = waitForWatching.runAsync(s1)
+        val waitCancelHandle = waitAndCancelWatching.runAsync(s1)
+
+        try Await.result(waitCancelHandle, FiniteDuration(3, TimeUnit.SECONDS))
+        finally println(s"Cancellation was executed ${waitCancelHandle.isCompleted}")
+        try Await.result(waitWatchingHandle, FiniteDuration(5, TimeUnit.SECONDS))
+        finally println(s"Watching task was completed ${waitWatchingHandle.isCompleted}")
+    }
+    ()
```

Comment:
It's not immediately clear to me how this test can fail, can you please explain?

Reply:
This test can fail if the watching handle doesn't return (as it was happening before), which means that Bloop is blocked.